crates/bcr-wallet-core/src/purse.rs

```diff
@@ -80,4 +137,160 @@ where
+        let Some(wlt) = self.wallets.lock().unwrap().get(idx).cloned() else {
+            return Err(Error::WalletNotFound(idx));
+        };
+        let summary = wlt.prepare_pay(input, now).await?;
+        let pref = PaymentReference {
+            payment_ref: summary.request_id,
+            wallet_idx: idx,
+        };
+        *self.current_payment.lock().unwrap() = Some(pref);
+        Ok(summary)
+    }
+
+    pub async fn pay(&self, p_id: Uuid, tstamp: u64) -> Result<TransactionId> {
+        let p_ref = self.current_payment.lock().unwrap().take();
+        let Some(pref) = p_ref else {
+            return Err(Error::NoPrepareRef(p_id));
+        };
+        if pref.payment_ref != p_id {
+            return Err(Error::NoPrepareRef(p_id));
+        }
+        let Some(wlt) = self.wallets.lock().unwrap().get(pref.wallet_idx).cloned() else {
+            return Err(Error::Internal(String::from(
+                "Wallet not found for payment",
+            )));
+        };
+        let txid = wlt.pay(p_id, &self.nostr_cl, &self.http_cl, tstamp).await?;
+        Ok(txid)
+    }
+
+    pub fn prepare_pay_request(
+        &self,
+        amount: Amount,
+        unit: Option<CurrencyUnit>,
+        description: Option<String>,
+    ) -> Result<cdk18::PaymentRequest> {
+        let mints = {
+            let wlts = self.wallets.lock().unwrap();
+            let mut mints = Vec::with_capacity(wlts.len());
+            for wlt in wlts.iter() {
+                mints.push(wlt.mint_url());
+            }
+            mints
+        };
+        let nostr_transport = cdk18::Transport {
+            _type: cdk18::TransportType::Nostr,
+            target: self.myself.to_bech32()?,
+            tags: Some(vec![vec![String::from("n"), String::from("17")]]),
+        };
+        let request = cdk18::PaymentRequest {
+            payment_id: Some(Uuid::new_v4().to_string()),
+            amount: Some(amount),
+            mints: Some(mints),
+            unit,
+            single_use: Some(true),
+            description,
+            nut10: None,
+            transports: Some(vec![nostr_transport]),
+        };
+        Ok(request)
+    }
+
+    pub async fn check_received_payment(
+        &self,
+        max_wait: core::time::Duration,
+        p_id: &str,
+        now: u64,
+    ) -> Result<Option<TransactionId>> {
+        let wlts = Arc::clone(&self.wallets);
+        let sub_id = self.nostr_subid.clone();
+        let cl = Arc::clone(&self.nostr_cl);
+        let tx_id: Arc<Mutex<Option<TransactionId>>> = Arc::new(Mutex::new(None));
+        let notif_handler = self.nostr_cl.handle_notifications(|notif| async {
+            let cloned = Arc::clone(&wlts);
+            let txid_recipient = Arc::clone(&tx_id);
+            let ok =
+                handle_notification(notif, &cl, &sub_id, cloned, p_id, now, txid_recipient).await?;
+            Ok(ok)
+        });
+        match tokio::time::timeout(max_wait, notif_handler).await {
```

Comment:
If you don't intend to listen for push notifications permanently you can just query for your message filter directly like:
```
client.fetch_events_from(
  relays (eg from config),
  filter (the filter as you have it for subscription),
  timeout,
)
// or:
client.fetch_events(filter, timeout) // using the clients relays
```

Reply:
Timeout is similar to a timeout of an http request. From the method docs `This is an **auto-closing subscription** and will be closed automatically on 'EOSE'`, where EOSE is a message from the relay saying end of stream.